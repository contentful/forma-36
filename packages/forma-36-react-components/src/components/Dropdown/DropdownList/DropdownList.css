@import 'settings/colors';
@import 'settings/typography';

.DropdownList {
  padding: 0;
  margin: 0;
  padding: calc(1rem * (14 / var(--font-base-default))) 0;
}

/* For multiple DropdownLists */
.DropdownList + .DropdownList {
  padding: calc(1rem * (7 / var(--font-base-default))) 0;
}

/* First DropdownList if multiple */
.DropdownList:first-child:not(:only-child) {
  padding-bottom: calc(1rem * (7 / var(--font-base-default)));
}

/* Last DropdownList if multiple */
.DropdownList:last-child:not(:only-child) {
  padding-bottom: calc(1rem * (14 / var(--font-base-default)));
}

.DropdownList--border-top {
  border-top: calc(1rem * (1 / var(--font-base-default))) solid
    var(--color-element-mid);
}

.DropdownList--border-bottom {
  border-bottom: calc(1rem * (1 / var(--font-base-default))) solid
    var(--color-element-mid);
}
