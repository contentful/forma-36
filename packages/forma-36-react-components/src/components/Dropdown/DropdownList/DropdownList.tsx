import React from 'react';
import cn from 'classnames';
import type * as CSS from 'csstype';

import cssStyles from './DropdownList.css';

export interface DropdownListProps {
  children: React.ReactNode;
  listRef?: React.RefObject<HTMLUListElement>;
  className?: string;
  testId?: string;
  border?: 'top' | 'bottom';
  maxHeight?: number | CSS.Property.MaxHeight;
  styles?: object;
}

export const DropdownList = ({
  className,
  border,
  maxHeight,
  testId,
  children,
  listRef,
  styles,
  ...otherProps
}: DropdownListProps) => {
  const classNames = cn(cssStyles['DropdownList'], className, {
    [cssStyles[`DropdownList--border-${border}`]]: border,
  });

  return (
    <ul
      ref={listRef}
      data-test-id={testId}
      role="menu"
      style={{
        maxHeight: maxHeight || 'auto',
        overflowY: maxHeight ? 'auto' : 'visible',
        ...styles,
      }}
      className={classNames}
      {...otherProps}
    >
      {children}
    </ul>
  );
};

DropdownList.defaultProps = {
  testId: 'cf-ui-dropdown-list',
};

export default DropdownList;
