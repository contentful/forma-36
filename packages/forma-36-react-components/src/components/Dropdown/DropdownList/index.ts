export * from './DropdownList';
export { default } from './DropdownList';
