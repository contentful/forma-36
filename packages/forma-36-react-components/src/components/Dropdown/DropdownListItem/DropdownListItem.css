@import 'settings/colors';
@import 'settings/typography';
@import 'settings/dimensions';

.DropdownListItem {
  display: inline-block;
  padding: var(--spacing-xs) var(--spacing-m);
  font-family: var(--font-stack-primary);
  font-size: var(--font-size-m);
  white-space: nowrap;
  display: block;
  position: relative;
  color: var(--color-text-mid);
  font-weight: var(--font-weight-normal);
  margin: 0;
  word-break: break-word;
  white-space: break-spaces;
  hyphens: auto;
}

.DropdownListItem__button:hover {
  background: var(--color-element-lightest);
}

.DropdownListItem__submenu-toggle {
  padding: 0px;
}

.DropdownListItem__button {
  composes: focus-outline--default from './../../../styles/settings/a11y.css';
  border: none;
  background: transparent;
  font-family: var(--font-stack-primary);
  width: 100%;
  font-size: var(--font-size-m);
  text-align: left;
  padding: 0;
  cursor: pointer;
  color: var(--color-text-mid);
  display: block;
  text-decoration: none;
}

a.DropdownListItem__button:link {
  color: var(--color-text-mid);
}

.DropdownListItem__button__inner-wrapper {
  display: block;
  padding: var(--spacing-xs) var(--spacing-m);
  line-height: 1.5;
}

.DropdownListItem--disabled,
.DropdownListItem--disabled button {
  background: var(--color-white);
  opacity: 0.5;
  cursor: not-allowed;

  &:hover {
    background: var(--color-white);
  }
}

.DropdownListItem--active,
.DropdownListItem--active:hover {
  background: var(--color-element-light);
}

.DropdownListItem--active .DropdownListItem__button:hover {
  background: var(--color-element-light);
  cursor: default;
}

.DropdownListItem--title {
  line-height: 1.8;
  padding-bottom: var(--spacing-xs);
}

.DropdownListItem--title,
.DropdownListItem--title:hover {
  color: var(--color-text-dark);
  background: var(--color-white);
  font-size: calc(1rem * (12 / var(--font-base-default)));
  font-weight: var(--font-weight-demi-bold);
  text-transform: uppercase;
  letter-spacing: 0.1rem;
}
