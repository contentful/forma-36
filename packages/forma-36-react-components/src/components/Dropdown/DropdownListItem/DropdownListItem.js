import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import TabFocusTrap from '../../TabFocusTrap/TabFocusTrap';
import styles from './DropdownListItem.css';

class DropdownListItem extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    children: PropTypes.node.isRequired,
    onClick: PropTypes.func,
    testId: PropTypes.string,
    onMouseDown: PropTypes.func,
    submenuToggleLabel: PropTypes.string,
    onFocus: PropTypes.func,
    onLeave: PropTypes.func,
    onEnter: PropTypes.func,
    isDisabled: PropTypes.bool,
    isActive: PropTypes.bool,
    isTitle: PropTypes.bool,
  };

  static defaultProps = {
    extraClassNames: undefined,
    submenuToggleLabel: undefined,
    onClick: undefined,
    testId: 'cf-ui-dropdown-list-item',
    onMouseDown: undefined,
    onFocus: undefined,
    onLeave: () => {},
    onEnter: () => {},
    isDisabled: false,
    isActive: false,
    isTitle: false,
  };

  renderSubmenuToggle = () => (
    <React.Fragment>
      <button
        type="button"
        data-test-id="cf-ui-dropdown-submenu-toggle"
        className={styles['DropdownListItem__toggle-button']}
        onClick={this.props.onClick}
        onMouseEnter={this.props.onEnter}
        onFocus={this.props.onEnter}
        onMouseLeave={this.props.onLeave}
        {...this.otherProps}
      >
        <TabFocusTrap
          extraClassNames={
            styles['DropdownListItem__toggle-button__inner-wrapper']
          }
        >
          {this.props.submenuToggleLabel}
        </TabFocusTrap>
      </button>
      {this.props.children}
    </React.Fragment>
  );

  renderListItem = () =>
    this.props.onClick || this.props.onMouseDown ? (
      <button
        type="button"
        data-test-id="cf-ui-dropdown-list-item-button"
        className={styles['DropdownListItem__toggle-button']}
        onClick={e => {
          if (!this.props.isDisabled && this.props.onClick) {
            this.props.onClick(e);
          }
        }}
        onMouseDown={e => {
          if (!this.props.isDisabled && this.props.onMouseDown) {
            this.props.onMouseDown(e);
          }
        }}
      >
        <TabFocusTrap
          extraClassNames={
            styles['DropdownListItem__toggle-button__inner-wrapper']
          }
        >
          {this.props.children}
        </TabFocusTrap>
      </button>
    ) : (
      this.props.children
    );

  render() {
    const {
      extraClassNames,
      isDisabled,
      testId,
      isActive,
      onClick,
      onMouseDown,
      submenuToggleLabel,
      isTitle,
    } = this.props;

    const classNames = cn(styles.DropdownListItem, extraClassNames, {
      [styles['DropdownListItem__submenu-toggle']]:
        submenuToggleLabel || onClick || onMouseDown,
      [styles['DropdownListItem--disabled']]: isDisabled,
      [styles['DropdownListItem--active']]: isActive,
      [styles['DropdownListItem--title']]: isTitle,
    });

    return (
      <li className={classNames} data-test-id={testId}>
        {submenuToggleLabel
          ? this.renderSubmenuToggle()
          : this.renderListItem()}
      </li>
    );
  }
}

export default DropdownListItem;
