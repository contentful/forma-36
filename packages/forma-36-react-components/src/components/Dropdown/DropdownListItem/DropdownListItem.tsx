import React, {
  forwardRef,
  MouseEventHandler,
  FocusEventHandler,
  useCallback,
} from 'react';
import cn from 'classnames';

import TabFocusTrap from '../../TabFocusTrap/TabFocusTrap';
import styles from './DropdownListItem.css';

export interface DropdownListItemProps
  extends React.HTMLAttributes<HTMLElement> {
  isDisabled?: boolean;
  listItemRef?: React.MutableRefObject<HTMLLIElement | null>;
  isActive?: boolean;
  isTitle?: boolean;
  children: React.ReactNode;
  onClick?: MouseEventHandler;
  onMouseDown?: MouseEventHandler;
  submenuToggleLabel?: string;
  href?: string;
  onFocus?: FocusEventHandler;
  onLeave?: MouseEventHandler;
  onEnter?: MouseEventHandler;
  className?: string;
  testId?: string;
}

export const DropdownListItem = forwardRef<HTMLElement, DropdownListItemProps>(
  function DropdownListItem(
    {
      isActive,
      isDisabled,
      isTitle,
      onClick,
      submenuToggleLabel,
      testId,
      ...props
    },
    refCallback,
  ) {
    // We're not dealing with React RefObjects but with useState (because we
    // want to re-render on all changes)
    const setReference = refCallback as React.Dispatch<
      React.SetStateAction<HTMLLIElement | null>
    >;

    const renderSubmenuToggle = useCallback(() => {
      const { onEnter, onLeave, onFocus, children, ...otherProps } = props;

      return (
        <React.Fragment>
          <button
            type="button"
            data-test-id="cf-ui-dropdown-submenu-toggle"
            className={styles['DropdownListItem__button']}
            onClick={onClick}
            onMouseEnter={onEnter}
            onFocus={onFocus}
            onMouseLeave={onLeave}
            {...otherProps}
          >
            <TabFocusTrap
              className={styles['DropdownListItem__button__inner-wrapper']}
            >
              {submenuToggleLabel}
            </TabFocusTrap>
          </button>
          {children}
        </React.Fragment>
      );
    }, [onClick, props]);

    const renderListItem = useCallback(() => {
      const { onMouseDown, href, children, listItemRef, ...otherProps } = props;

      const isClickable = onClick || onMouseDown || href;

      if (isClickable) {
        const Element = href ? 'a' : 'button';

        const buttonProps = {
          disabled: isDisabled,
          'aria-disabled': isDisabled,
        };

        const linkProps = {
          href,
        };

        return (
          <Element
            onClick={!isDisabled && onClick !== undefined ? onClick : undefined}
            onMouseDown={
              !isDisabled && onMouseDown !== undefined ? onMouseDown : undefined
            }
            type="button"
            {...(href ? linkProps : buttonProps)}
            {...otherProps}
            className={styles['DropdownListItem__button']}
            data-test-id="cf-ui-dropdown-list-item-button"
          >
            <TabFocusTrap
              className={styles['DropdownListItem__button__inner-wrapper']}
            >
              {children}
            </TabFocusTrap>
          </Element>
        );
      }

      return children;
    }, [onClick, props]);

    const { className, listItemRef, onMouseDown, href } = props;

    const classNames = cn(styles['DropdownListItem'], className, {
      [styles['DropdownListItem__submenu-toggle']]:
        submenuToggleLabel || onClick || onMouseDown || href,
      [styles['DropdownListItem--disabled']]: isDisabled,
      [styles['DropdownListItem--active']]: isActive,
      [styles['DropdownListItem--title']]: isTitle,
    });

    return (
      <li
        className={classNames}
        data-test-id={testId}
        role="menuitem"
        ref={(node) => {
          if (setReference) {
            setReference(node);
          }

          if (listItemRef) {
            listItemRef.current = node;
          }
        }}
      >
        {submenuToggleLabel ? renderSubmenuToggle() : renderListItem()}
      </li>
    );
  },
);

DropdownListItem.displayName = 'DropdownListItem';

DropdownListItem.defaultProps = {
  testId: 'cf-ui-dropdown-list-item',
  isDisabled: false,
  isActive: false,
  isTitle: false,
};

export default DropdownListItem;
