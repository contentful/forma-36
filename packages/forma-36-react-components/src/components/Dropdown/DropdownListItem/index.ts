export * from './DropdownListItem';
export { default } from './DropdownListItem';
