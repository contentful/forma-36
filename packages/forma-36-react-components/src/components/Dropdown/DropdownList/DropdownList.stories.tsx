import React from 'react';

import { DropdownList } from './DropdownList';
import { DropdownListItem } from '../DropdownListItem/DropdownListItem';

export default {
  title: 'Components/Dropdown/DropdownList',
  component: DropdownList,
  parameters: {
    propTypes: [DropdownList['__docgenInfo']],
  },
  argTypes: {
    className: { control: { disable: true } },
    testId: { control: { disable: true } },
  },
};

export const Basic = () => (
  <DropdownList>
    <DropdownListItem isTitle>Title Text</DropdownListItem>
    <DropdownListItem isDisabled>Block entry disabled</DropdownListItem>
  </DropdownList>
);
