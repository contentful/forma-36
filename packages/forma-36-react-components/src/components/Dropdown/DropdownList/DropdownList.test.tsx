import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import DropdownList from './DropdownList';
import DropdownListItem from '../DropdownListItem/DropdownListItem';

it('renders the component', () => {
  const output = shallow(
    <DropdownList>
      <DropdownListItem>List Item</DropdownListItem>
    </DropdownList>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <DropdownList className="my-extra-class">
      <DropdownListItem>List Item</DropdownListItem>
    </DropdownList>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with a border attribute', () => {
  const output = shallow(
    <DropdownList border="top">
      <DropdownListItem>List Item</DropdownListItem>
    </DropdownList>,
  );

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(
    <DropdownList>
      <DropdownListItem>entry</DropdownListItem>
    </DropdownList>,
  ).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
