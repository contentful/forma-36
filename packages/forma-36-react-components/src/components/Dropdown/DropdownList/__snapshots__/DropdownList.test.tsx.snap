// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<ul
  class="DropdownList"
  data-test-id="cf-ui-dropdown-list"
  role="menu"
  style="max-height: auto; overflow-y: visible;"
>
  <li
    class="DropdownListItem"
    data-test-id="cf-ui-dropdown-list-item"
    role="menuitem"
  >
    List Item
  </li>
</ul>
`;

exports[`renders the component with a border attribute 1`] = `
<ul
  class="DropdownList DropdownList--border-top"
  data-test-id="cf-ui-dropdown-list"
  role="menu"
  style="max-height: auto; overflow-y: visible;"
>
  <li
    class="DropdownListItem"
    data-test-id="cf-ui-dropdown-list-item"
    role="menuitem"
  >
    List Item
  </li>
</ul>
`;

exports[`renders the component with an additional class name 1`] = `
<ul
  class="DropdownList my-extra-class"
  data-test-id="cf-ui-dropdown-list"
  role="menu"
  style="max-height: auto; overflow-y: visible;"
>
  <li
    class="DropdownListItem"
    data-test-id="cf-ui-dropdown-list-item"
    role="menuitem"
  >
    List Item
  </li>
</ul>
`;
